graylog2-server/src/main/java/org/graylog2/configuration/IndexerDiscoveryProvider.java

```diff
@@ -68,16 +69,17 @@ public IndexerDiscoveryProvider(
             @Named("datanode_startup_connection_delay") Duration delayBetweenAttempts,
             PreflightConfigService preflightConfigService,
             NodeService<DataNodeDto> nodeService,
-            GraylogCertificateProvisioner graylogCertificateProvisioner) {
+            Set<IndexerDiscoveryListener> indexerDiscoveryListeners) {
```

Comment:
Do you think we will ever have more listeners?
Right now the two listeners have very distinct functionalities at different stages. I think we could also just split that into two services.

Reply:
It felt like a good abstraction, a hook into the discovery process, before start and during each retry. 

I am happy that we moved the certificate provisioning out of this part of the code and made it pluggable.  That was definitely too tight coupling and deserved a refactoring 

The discovery lifecycle here is the common functionality for listeners, even if they are doing completely different things. How would you suggest to splitting it?

Reply:
Nothing wrong with doing it the way it is implemented. 
It only feels a bit awkward that we have two implementations of the listener that are triggered at very specific lifecycle stages and I am not sure if we will ever have more. So the abstraction might have been unnecessary even if it would have come with tighter coupling.
But I can definitely live with how it is implemented. Less work if we ever need to extend on it.